Support indices contiguity per batch and non-contiguous values in sparse compressed tensors

pearu opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91331
* #90849
* __->__ #91243



cc @nikitaved @cpuhrsch @amjames @bhosmer

pearu committed:
Support non-contiguous inputs in validation of sparse compressed tensors

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -1,17 +1,19 @@
 #include <ATen/native/TensorIterator.h>
-#include <ATen/Dispatch.h>
 #include <ATen/native/sparse/Macros.h>
+#include <iostream>
```
to be removed

pearu replied:
fixed

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const Tensor batch_shape = at::tensor(batch_dims, idx.options());
+    const Tensor batch_idx_stride =
+        at::tensor(idx.strides().slice(0, idx.dim() - 1), idx.options());
```
Shape and strides are int64. Let us probably keep it that way for these newly created tensors.

pearu replied:
Resolved.

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
+    const Tensor batch_shape = at::tensor(batch_dims, idx.options());
+    const Tensor batch_idx_stride =
+        at::tensor(idx.strides().slice(0, idx.dim() - 1), idx.options());
+    const Tensor all_idx_offsets = at::arange(
+        batch_idx_stride.mul(batch_shape.sub(1)).sum().add(1).item(),
+        idx.options());
+    const Tensor idx_offsets =
+        all_idx_offsets.reshape({-1, 1})
+            .div(batch_idx_stride, "floor") // TODO: support zero strides
+            .remainder(batch_shape)
+            .mul(batch_idx_stride)
+            .sum(1)
+            .eq(all_idx_offsets)
+            .nonzero()
+            .to(idx.dtype())
+            .view(batch_dims)
+            .unsqueeze_(-1);
```
These ops could be done on the CPU, and then moved to the GPU if needed.

nikitaved replied:
Maybe, if that is of use, there is a function `make_offset_calculator` that allows to map a linear index to a multi-dim offset.

pearu replied:
The computation of idx offsets is moved to the kernel.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -174,6 +184,15 @@ void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_ind
               "values must have dimensionality > sum of batch and block dimensionalities (=",
               batch_ndim, " + ", block_ndim, ") but got ", values.dim());
+
+  // 3.5
+  TORCH_CHECK(plain_indices.stride(plain_indices.dim() - 1) == 1,
+              "expected ", plain_indices_name, " to be a contiguous tensor per batch");
+
```
Will use `plain_indices.stride(-1)`, thanks for the hint!

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -864,10 +864,10 @@ void add_out_dense_sparse_csr_cpu(
   resultBuffer = resultBuffer.view({-1, out.size(-2), out.size(-1)});
-  auto src_crow_indices = src.crow_indices().view({-1, src.crow_indices().size(-1)});
-  auto src_col_indices = src.col_indices().view({-1, src.col_indices().size(-1)});
+  auto src_crow_indices = src.crow_indices().reshape({-1, src.crow_indices().size(-1)});
+  auto src_col_indices = src.col_indices().reshape({-1, src.col_indices().size(-1)});
```
The CPP version uses just `reshape` while the CUDA version is doing `reshape + contiguous`. Either way, if `reshape` is called (or `flatten`), the batch dimensions become regular, and there is no need in sophisticated maps from a linear index to a multi-dim offset.
It seems like dropping `reshape` is best for performance, and we could use `make_offset_calculator` to keep the CPU and GPU kernel relatively intact. Otherwise it is best to re-implement these kernels and use `TensorIterator.for_each`.

pearu replied:
> The CPP version uses just reshape while the CUDA version is doing reshape + contiguous

This is due to CPU version using accessor indexing while the CUDA version uses pointer indexing.

The rest of the comments are OT to this PR but I agree that the CPU and GPU kernels could be unified here.

nikitaved replied:
Why are they different? They use the same code...

nikitaved replied:
Sorry, @pearu, I thought these are inside invariant checks, but they turn out to be in the `add` op. Apologies for the noise...

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
+        batch_idx_stride.mul(batch_shape.sub(1)).sum().add(1).item(),
+        idx.options());
+    const Tensor idx_offsets =
+        all_idx_offsets.reshape({-1, 1})
+            .div(batch_idx_stride, "floor") // TODO: support zero strides
+            .remainder(batch_shape)
```
Is it easy to fix? Do we get an error message and not a segfault if there are zero strides?

pearu replied:
`make_offset_calculator` appears to be CUDA specific but there seems to be analogous code for the CPU. I'll investigate that..

Re zero strides: I think we should support it because otherwise, test_bmm needs to use an unorthodox way of constructing CSR tensors. I suggest leaving it as a follow-up PR within this ghstack as in addition to generalizing the above, the generate_simple_inputs need to be updated accordingly and it is unpredictable how other existing tests behave on zero-strided inputs.

nikitaved replied:
Then, I guess, it would be great to just error out, and then remove it in a follow-up as to avoid any potential segfaults...

pearu replied:
Fixed, the latest commit supports zero strides.

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +272,83 @@ void _validate_compressed_sparse_indices_kernel(
+    const auto idx_ndims = idx.dim();
+    const auto idx_sizes = idx.sizes();
+    const auto idx_strides = idx.strides();
+    const auto idx_contiguous_strides = c10::contiguous_strides(idx_sizes);
+    std::array<int64_t, 3 * MAX_DIMS> idx_sizes_and_strides;
```
nit: having 3 separate arrays might be a bit easier to understand.

pearu replied:
Fixed.

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +272,83 @@ void _validate_compressed_sparse_indices_kernel(
+                // assuming idx contiguity per batch:
+                const int64_t batch_idx_ = batch_idx * sizes[idx_ndims - 1];
+                for (int i = 0; i < idx_ndims - 1; i++) {
+                  idx_offset +=
+                      ((batch_idx_ / contiguous_strides[i]) % sizes[i]) *
+                      strides[i];
```
Do we really need `% sizes[i]`?

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +272,75 @@ void _validate_compressed_sparse_indices_kernel(
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const auto idx_ndims = idx.dim();
+    std::array<int64_t, MAX_DIMS> idx_sizes, idx_strides;
+    for (int i = 0; i < idx_ndims; i++) {
```
We should probably error out if `idx_ndims > MAX_DIMS`, otherwise it is a UB.

pearu replied:
Fixed: sizes and strides are now passed to the kernel via `Tensor` instance and we don't depend on `MAX_DIMS` anymore.

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -241,52 +272,75 @@ void _validate_compressed_sparse_indices_kernel(
+                // between different threads.
+                int64_t idx_offset = 0;
+                // assuming idx contiguity per batch:
+                int64_t tmp = batch_idx * idx_sizes[idx_ndims - 1];
+                for (int i = idx_ndims - 1; i >= 0; i--) {
+                  idx_offset += (tmp % idx_sizes[i]) * idx_strides[i];
+                  tmp /= idx_sizes[i];
+                }
```
Division and mod are slow on CUDA, especially compared to all other ops in the kernel. I am fine with that for now, but it would not hurt to optimize this case for contiguous inputs.

pearu replied:
I can eliminate `mod` here but overall this feels like an over-optimization considering that the invariants validation can be disabled if maximum performance is required.

pearu replied:
Eliminated `mod`.

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -276,11 +269,9 @@ void _validate_compressed_sparse_indices_kernel(
-      idx_strides[i] = idx.stride(i);
-    }
+
+    Tensor idx_sizes_ = at::tensor(idx.sizes(), idx.options().dtype(kLong));
+    Tensor idx_strides_ = at::tensor(idx.strides(), idx.options().dtype(kLong));
```
minor nit: we can create a buffer tensor on the cpu, copy sizes and strides into it, move it to CUDA, then `unbind` it into two separate CUDA tensors for only a single data transfer. But maybe it is an over-optimization...

pearu replied:
Done.

nikitaved commented on aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h:
```diff
@@ -322,8 +316,9 @@ void _validate_compressed_sparse_indices_kernel(
+                  int64_t div = tmp / idx_sizes[i];
+                  idx_offset += (tmp - div * idx_sizes[i]) * idx_strides[i];
+                  tmp = div;
                 }
```
👍

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2709,6 +2714,21 @@ def generate_values(base, densesize):
+            assert dim >= 0 and dim < t.ndim
+            step = max(2, offset + 1)
+            tmp = torch.zeros((*t.shape[:dim], t.shape[dim] * step, *t.shape[dim + 1:]), dtype=t.dtype, device=t.device)
+            tmp[(*((slice(None),) * dim), slice(offset, None, step))].copy_(t)
```
nit: `tmp[..., slice(...)]` should work as well.

pearu replied:
Thanks! I'll apply this in some follow-up.

pearu replied:
Using `Ellipsis` does not work because `offset::step` is not the last dimension.

nikitaved reviewed (approved):
LGTM. Thank you, @pearu !

pearu commented:
@pytorchbot merge

pearu commented:
@YXIE14, @842974287, @terrychenism, @fuqianz, or @Xirider , could you approve this PR? Thanks!

pearu committed:
Update on "Support indices contiguity per batch and non-contiguous values in sparse compressed tensors"

Fixes https://github.com/pytorch/pytorch/issues/91062

With this PR, all reported failures in https://github.com/pytorch/pytorch/pull/90849 are resolved (modulo test_bmm that uses an unorthodox way to construct a batch CSR tensor).




cc nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

amjames commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge
